File: src/main/java/com/nerdearla/workshop/paymentMethod/PaymentMethodService.java
package com.nerdearla.workshop.paymentMethod;

import com.nerdearla.workshop.payment.service.model.PaymentMethodData;
import com.nerdearla.workshop.paymentMethod.client.PaymentMethodClient;
import com.nerdearla.workshop.paymentMethod.validator.PaymentMethodValidator;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

@Service
public class PaymentMethodService {

    private static final Logger LOGGER = LoggerFactory.getLogger(PaymentMethodService.class);

    private final PaymentMethodClient client;
    private final PaymentMethodValidator validator;

    public PaymentMethodService(
            PaymentMethodClient client,
            PaymentMethodValidator validator) {
        this.client = client;
        this.validator = validator;
    }

    public BuyerPaymentMethod getBy(String buyerId, PaymentMethodData paymentMethodData) {
        BuyerPaymentMethod buyerPaymentMethod = doGet(buyerId, paymentMethodData.getToken());
        validate(buyerPaymentMethod, paymentMethodData);
        return buyerPaymentMethod;
    }

    private BuyerPaymentMethod doGet(String buyerId, String token) {
        BuyerPaymentMethod buyerPaymentMethod = client.get(buyerId, token);
        LOGGER.info("payment method found: {}", buyerPaymentMethod.toString());
        return buyerPaymentMethod;
    }

    private void validate(BuyerPaymentMethod buyerPaymentMethod, PaymentMethodData paymentMethodData) {
        validator.validate(buyerPaymentMethod, paymentMethodData);
        LOGGER.info("payment method {} validated successfully", buyerPaymentMethod.getId());
    }
}

File: src/main/kotlin/com/nerdearla/workshop/paymentMethod/PaymentMethodService.kt
package com.nerdearla.workshop.paymentMethod

import com.nerdearla.workshop.payment.service.model.PaymentMethodData
import com.nerdearla.workshop.paymentMethod.client.PaymentMethodClient
import com.nerdearla.workshop.paymentMethod.validator.PaymentMethodValidator
import com.nerdearla.workshop.shared.utils.CompanionLogger
import org.springframework.stereotype.Service

@Service
class PaymentMethodService(
    private val client: PaymentMethodClient,
    private val validator: PaymentMethodValidator
) {
    fun getBy(buyerId: String, paymentMethodData: PaymentMethodData): BuyerPaymentMethod =
        doGet(buyerId, paymentMethodData.token)
            .also { it.validateAgainst(paymentMethodData) }

    private fun doGet(buyerId: String, token: String) =
        client.getBy(buyerId, token)
            .log { info("payment method found: {}", it) }

    private fun BuyerPaymentMethod.validateAgainst(paymentMethodData: PaymentMethodData) =
        validator.validate(this, paymentMethodData)
            .log { info("payment {} validated", id) }

    companion object : CompanionLogger()
}


